Index: python2.6-2.6.8/Lib/test/test_socket.py
===================================================================
--- python2.6-2.6.8.orig/Lib/test/test_socket.py	2014-06-04 11:31:07.138253680 +0200
+++ python2.6-2.6.8/Lib/test/test_socket.py	2014-06-04 12:11:17.004667597 +0200
@@ -1239,6 +1239,16 @@ class BufferIOTest(SocketConnectedTest):
             buf = buffer(MSG)
         self.serv_conn.send(buf)
 
+    def testRecvFromIntoSmallBuffer(self):
+        # See issue #20246.
+        buf = array.array('c', ' '*8)
+        self.assertRaises(ValueError, self.cli_conn.recvfrom_into, buf, 1024)
+
+    def _testRecvFromIntoSmallBuffer(self):
+        with test_support._check_py3k_warnings():
+            buf = buffer(MSG*2048)
+        self.serv_conn.send(buf)
+
 
 TIPC_STYPE = 2000
 TIPC_LOWER = 200
Index: python2.6-2.6.8/Modules/socketmodule.c
===================================================================
--- python2.6-2.6.8.orig/Modules/socketmodule.c	2014-06-04 12:08:15.835624444 +0200
+++ python2.6-2.6.8/Modules/socketmodule.c	2014-06-04 12:08:18.015637091 +0200
@@ -2625,6 +2625,10 @@ sock_recvfrom_into(PySocketSockObject *s
     if (recvlen == 0) {
         /* If nbytes was not specified, use the buffer's length */
         recvlen = buflen;
+    } else if (recvlen > buflen) {
+        PyErr_SetString(PyExc_ValueError,
+                        "nbytes is greater than the length of the buffer");
+        return NULL;
     }
 
     readlen = sock_recvfrom_guts(s, buf, recvlen, flags, &addr);
